pkg/object/gatewaycontroller/k8s.go

```diff
@@ -0,0 +1,490 @@
+/*
+ * Copyright (c) 2017, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package gatewaycontroller
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/megaease/easegress/v2/pkg/logger"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/discovery"
+	corev1 "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/informers/internalinterfaces"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+
+	apicorev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+
+	kinformers "k8s.io/client-go/informers"
+	kclientset "k8s.io/client-go/kubernetes"
+
+	gwapis "sigs.k8s.io/gateway-api/apis/v1beta1"
+	gwclientset "sigs.k8s.io/gateway-api/pkg/client/clientset/versioned"
+	gwinformers "sigs.k8s.io/gateway-api/pkg/client/informers/externalversions"
+	gwinformerapis "sigs.k8s.io/gateway-api/pkg/client/informers/externalversions/apis/v1beta1"
+)
+
+const (
+	resyncPeriod = 10 * time.Minute
+)
+
+func isResourceChanged(oldObj, newObj interface{}) bool {
+	if oldObj == nil || newObj == nil {
+		return true
+	}
+
+	if oldObj.(metav1.Object).GetResourceVersion() == newObj.(metav1.Object).GetResourceVersion() {
+		return false
+	}
+
+	if ep, ok := oldObj.(*apicorev1.Endpoints); ok {
+		return isEndpointsChanged(ep, newObj.(*apicorev1.Endpoints))
+	}
+
+	return true
+}
+
+func isEndpointsChanged(a, b *apicorev1.Endpoints) bool {
+	if len(a.Subsets) != len(b.Subsets) {
+		return true
+	}
+
+	for i, sa := range a.Subsets {
+		sb := b.Subsets[i]
+		if isSubsetsChanged(sa, sb) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func isSubsetsChanged(a, b apicorev1.EndpointSubset) bool {
```

Comment:
Why `reflect.DeepEqual` can not cover this function.

Reply:
`DeepEqual` can be used here.
But for performance consideration, explicit comparsion is used.